### ISSUE TITLE
ImportError: No module named nmtwizard.framework

### ISSUE THREAD

**Issue by paulkp** [OpenNMT/nmt-wizard-docker#145]
I've installed this repo and nvidia drivers on an AWS instance.
Now trying to run:

`python frameworks/opennmt_py/entrypoint.py --model_storage /tmp/saved-models --config frameworks/opennmt_py/config/train_ende_example.json train`

I get:

> Traceback (most recent call last):
>   File "frameworks/opennmt_py/entrypoint.py", line 7, in <module>
>     from nmtwizard.framework import Framework
> ImportError: No module named nmtwizard.framework

and I HAVE installed requirements.txt at 2 levels.

But a module named nmtwizard is not included in any requirements.txt.
And I get 'doesn't exist' if I try to pip install nmtwizard.

So what's the story?
I'm trying to follow this repo's README.

**Comment by guillaumekln** [comment]
It seems the `nmtwizard` directory is not in your PYTHONPATH. Can you try using the Docker images instead?

**Comment by paulkp** [comment]
> It seems the `nmtwizard` directory is not in your PYTHONPATH. Can you try using the Docker images instead?

Yes @guillaumekln , thanks, but: 
Q. exactly how do we use/install the Docker image? I think my guy has just git pulled the code . .

(I can see the directions at the end of the readme but that seems to be how to run a training session)

**Comment by guillaumekln** [comment]
For each new tag on this repository, we push Docker images in the [`nmtwizard` organization](https://hub.docker.com/u/nmtwizard). Try for example:

```bash
docker run --rm nmtwizard/opennmt-py:1.23.0 -h
```

**Comment by paulkp** [comment]
> For each new tag on this repository, we push Docker images in the [`nmtwizard` organization](https://hub.docker.com/u/nmtwizard). Try for example:
> 
> ```shell
> docker run --rm nmtwizard/opennmt-py:1.23.0 -h
> ```

Thnx @guillaumekln, but 
Q1. How was I supposed to know to do that?
Q2. Why not provide Dockerfiles?
Q3. What is going on in that command? There is no file 'opennmt-py:1.23.0' in the repo's nmtwizard folder??

**Comment by guillaumekln** [comment]
1. This is the first command in the [readme](https://github.com/OpenNMT/nmt-wizard-docker/#overview). It displays the options exposed by the Docker image.
2. Do you mean how to run Docker images? See https://docs.docker.com/engine/reference/run/
3. There should be no need to clone the Git repository unless for development.

**Comment by paulkp** [comment]
@guillaumekln, I'm no Docker expert, but I've built and run Docker images before.
From a Dockerfile.
In this case you're saying I can install the OpenNMT Wizard Docker image by:

1. Installing Docker?
2. eg Typing:
docker run nmtwizard/opennmt-tf -h
or, in my case:
docker run nmtwizard/opennmt-py -h

and that should go and grab an image, presumably from Docker?
Or show me some 'options' you said?

**Comment by guillaumekln** [comment]
The 2 steps are correct. Docker automatically downloads images that are not built locally when using `docker run`. If you prefer, you can still build the image yourself:

```bash
docker build -t nmtwizard/opennmt-py -f frameworks/opennmt_py/Dockerfile .
```

or download it without running it:

```bash
docker pull nmtwizard/opennmt-py
```

The `-h` option should print the options available for this image. Here is the output I have:

```text
$ docker run nmtwizard/opennmt-py:latest -h
Unable to find image 'nmtwizard/opennmt-py:latest' locally
latest: Pulling from nmtwizard/opennmt-py
Digest: sha256:64ae81a3630132882922103c0f00fd8d903f261aea9345f07ab16b2fdcea1a0e
Status: Downloaded newer image for nmtwizard/opennmt-py:latest
usage: entrypoint.py [-h] [-s STORAGE_CONFIG] [-t TASK_ID] [-i IMAGE]
                     [-b BEAT_URL] [-bi BEAT_INTERVAL]
                     [--statistics_url STATISTICS_URL] [-ms MODEL_STORAGE]
                     [-msr MODEL_STORAGE_READ] [-msw MODEL_STORAGE_WRITE]
                     [-c CONFIG] [-m MODEL] [-g GPUID] [--no_push]
                     {train,trans,release,serve,preprocess,buildvocab} ...

positional arguments:
  {train,trans,release,serve,preprocess,buildvocab}
                        Run type
    train               Run a training.
    trans               Run a translation.
    release             Release a model for serving.
    serve               Serve a model.
    preprocess          Sample and preprocess corpus.
    buildvocab          Build vocabularies.

optional arguments:
  -h, --help            show this help message and exit
  -s STORAGE_CONFIG, --storage_config STORAGE_CONFIG
                        Configuration of available storages as a file or a
                        JSON string. Setting "-" will read from the standard
                        input.
  -t TASK_ID, --task_id TASK_ID
                        Identifier of this run.
  -i IMAGE, --image IMAGE
                        Full URL (registry/image:tag) of the image used for
                        this run.
  -b BEAT_URL, --beat_url BEAT_URL
                        Endpoint that listens to beat requests (push
                        notifications of activity).
  -bi BEAT_INTERVAL, --beat_interval BEAT_INTERVAL
                        Interval of beat requests in seconds.
  --statistics_url STATISTICS_URL
                        Endpoint that listens to statistics summaries
                        generated at the end of the execution
  -ms MODEL_STORAGE, --model_storage MODEL_STORAGE
                        Model storage in the form <storage_id>:[<path>].
  -msr MODEL_STORAGE_READ, --model_storage_read MODEL_STORAGE_READ
                        Model storage to read from, in the form
                        <storage_id>:[<path>] (defaults to model_storage).
  -msw MODEL_STORAGE_WRITE, --model_storage_write MODEL_STORAGE_WRITE
                        Model storage to write to, in the form
                        <storage_id>:[<path>] (defaults to model_storage).
  -c CONFIG, --config CONFIG
                        Configuration as a file or a JSON string. Setting "-"
                        will read from the standard input.
  -m MODEL, --model MODEL
                        Model to load.
  -g GPUID, --gpuid GPUID
                        Comma-separated list of 1-indexed GPU identifiers (0
                        for CPU).
  --no_push             Do not push model.
```

**Comment by paulkp** [comment]
@guillaumekln 
Q1. Can you actually tell me what steps to do to get OPenNMT WIzard installed and running on AWS?
Q2. Does it include the frameworks or not? Or do I need to install them separately?

**Comment by guillaumekln** [comment]
> Just remember I successfully got opennmt-py running last month. It was fine. But your wizard is like nothing I or my colleagues have come across. We jsut don't know what it is or how to install it or how to use it.

What are you expectations in using this project (nmt-wizard-docker)? Can you describe what are the issues you have currently and how this project would resolve them?

The readme states "The aim of this project is to encapsulate training frameworks in Docker containers and expose a standardized interface". The nmt-wizard-docker project is nothing more than some Python scripts around training/translation frameworks. If you don't need this abstraction and are fine in using OpenNMT-py directly, there is probably no reason to use this project.

**Comment by paulkp** [comment]
@guillaumekln 
After testing it, I wanted to run multiple trainings for multiple language pairs with multiple (hyper)parameters. And compare to competitors using the wrappers to GNMT etc. Everything that the wizard promises.
But I first have to get it working. And understand WHAT it is.
If only there was a 'how to' install it and a 'how to' use it with REAL examples.

**Comment by guillaumekln** [comment]
It seems to me you'll have an easier time using OpenNMT-py directly. You could build the workflow that works for you around it.

**Comment by gussmith** [comment]
> @guillaumekln, I'm no Docker expert, but I've built and run Docker images before.
> From a Dockerfile.
> In this case you're saying I can install the OpenNMT Wizard Docker image by:
> 
> 1. Installing Docker?
> 2. eg Typing:
>    `docker run nmtwizard/opennmt-tf -h`
>    or, in my case:
>    `docker run nmtwizard/opennmt-py -h`
> 
> and that should go and grab an image, presumably from Docker?
> Or show me some 'options' you said?
> 
> (BTW, my guy said that's what he did, but I still get the error
> `ImportError: No module named nmtwizard.framework when I try to train`)

You have to follow the instructions here to pull an image from the Docker repo automatically, including setting up the Docker's official GPG key: 
https://docs.docker.com/install/linux/docker-ce/ubuntu/

**Comment by gussmith** [comment]
> It seems the `nmtwizard` directory is not in your PYTHONPATH. Can you try using the Docker images instead?

@guillaumekln
There is a lack of context in the instructions making it hard to follow how to use what you are providing. 

I have the docker image 
$ docker images
REPOSITORY                  TAG                                  IMAGE ID            CREATED             SIZE
nmtwizard/opennmt-py        latest                               d8ad13e2a035        7 days ago          4.34GB

Now, the ReadMe: 
https://github.com/OpenNMT/nmt-wizard-docker/tree/master/frameworks/opennmt_py

1. Install dependencies:
but where? 
I assumed like @paulkp that it was on the **local** machine. 
Is it on the local machine? 
Or is it in the container after running the docker image? 
But then, why would we have to install anything in the docker container/image? 

So if on the local machine, and I download the files in 
https://github.com/OpenNMT/nmt-wizard-docker/tree/master/frameworks/opennmt_py
in the directories on the local machine I put: 
in directory frameworks/opennmt_py/: requirements.py and entrypoint.py
in directory frameworks/opennmt_py/config: train_ende_example.json

Then I imagine that on the local machine I run: 
python frameworks/opennmt_py/entrypoint.py --model_storage /tmp/saved-models \
    --config frameworks/opennmt_py/config/train_ende_example.json train

**and then I assume that the python script will call the docker image 
nmtwizard/opennmt-py internally somehow (?)** 

and that it should work, but the requirements.txt is missing: 
six 
requests 
that entrypoint.py uses on the local machine. 

then we are back to @paulkp 's original point: 
ModuleNotFoundError: No module named 'nmtwizard'

If it is meant to be inside the container, then how to run the container in interactive mode?
docker run --gpus all -it --rm --shm-size=1g --ulimit memlock=-1 nmtwizard/opennmt-py
bring the error: 
entrypoint.py: error: at least one of --config or --model options must be set

Then how does one combine the general readme to use the docker : 
https://github.com/OpenNMT/nmt-wizard-docker
(which is really difficult to follow)
with the python command: 
https://github.com/OpenNMT/nmt-wizard-docker/tree/master/frameworks/opennmt_py

python frameworks/opennmt_py/entrypoint.py --model_storage /tmp/saved-models \
    --config frameworks/opennmt_py/config/train_ende_example.json train

It is not clear to me. 
Can you please provide a simple example of how it is meant to be used? 

I have the general idea, that some config file, model file must be set on the local machine, then the paths must be set, then one should call docker run and that the container will train the model and provide results in some local directory. 
Is that right?
Again, a simple trivial example, would really help demystify a lot. 
Thank you.

**Comment by guillaumekln** [comment]
Sorry for the lack of clear instructions. We did not make enough effort on this side.

I completely revisited the main README and added an example. @gussmith Could you give a look and see if it makes things clearer?

I think part of the confusion came from the instructions in `frameworks/*/README.md` which describe how to run the script **without** Docker. As it is not the recommended usage, I removed these instructions entirely.

**Comment by gussmith** [comment]
@guillaumekln Thanks! Quite a bit clearer. I have to become familiar with the OpenNMT framework at the same time. I will give it a go. 
I basically want to run a novel model written for OpenNMT that is defined by a model.py, and other files: dataset.py, train.py and a predict.py. 
It may be easier to learn how to run that in OpenNMT and then learn to adapt it to the configuration file and other requirements for nmt-wizard-docker (?).

**Comment by guillaumekln** [comment]
Yes, when it comes to experimentation and research you should probably start with OpenNMT-py or OpenNMT-tf directly.

This project is mostly production oriented and is useful when you need to industrialize model training and evaluation.
